File: ios-app/CalendarioCiclismo/Services/OfflineManager.swift
import Foundation
import Network
import WidgetKit

/// Monitor de conectividad de red. Observable desde SwiftUI para decidir si
/// mostrar modales de "sin conexión" al tocar enlaces/assets.
@MainActor
@Observable
final class NetworkMonitor {
    static let shared = NetworkMonitor()

    /// `true` cuando hay una ruta de red satisfactoria (wifi, celular, ethernet).
    var isOnline: Bool = true

    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "app.calendariociclismo.network-monitor")

    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            let online = path.status == .satisfied
            Task { @MainActor in
                self?.isOnline = online
            }
        }
        monitor.start(queue: queue)
    }

    deinit {
        monitor.cancel()
    }
}

/// Gestiona el modo sin conexión: estado, descarga proactiva y purgado de datos.
/// Descarga diariamente los próximos 14 días, el mes actual y siguiente,
/// y la temporada completa para uso offline.
@MainActor
@Observable
final class OfflineManager {
    static let shared = OfflineManager()

    // MARK: - Estado público

    /// Indica si el modo offline está activado.
    var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: "offline_mode_enabled") }
        set { UserDefaults.standard.set(newValue, forKey: "offline_mode_enabled") }
    }

    /// Indica si el usuario ha completado el onboarding de modo offline.
    var hasCompletedOnboarding: Bool {
        get { UserDefaults.standard.bool(forKey: "offline_onboarding_completed") }
        set { UserDefaults.standard.set(newValue, forKey: "offline_onboarding_completed") }
    }

    /// Indica si se está realizando una sincronización.
    var isSyncing = false

    /// Progreso de la sincronización (0.0 a 1.0).
    var syncProgress: Double = 0

    /// Texto descriptivo del paso actual de sincronización.
    var syncStatusText: String?

    /// Fecha de la última sincronización completada.
    var lastSyncDate: Date? {
        get {
            let interval = UserDefaults.standard.double(forKey: "offline_last_sync")
            return interval > 0 ? Date(timeIntervalSince1970: interval) : nil
        }
        set {
            UserDefaults.standard.set(newValue?.timeIntervalSince1970 ?? 0, forKey: "offline_last_sync")
        }
    }

    /// Texto legible con la fecha de última sincronización.
    var lastSyncLabel: String? {
        guard let date = lastSyncDate else { return nil }
        let seconds = Date().timeIntervalSince(date)
        let minutes = Int(seconds) / 60
        if minutes < 1 { return "Hace un momento" }
        if minutes < 60 { return "Hace \(minutes) min" }
        let hours = minutes / 60
        if hours < 24 { return "Hace \(hours) h" }
        let days = hours / 24
        return "Hace \(days) d"
    }

    /// Indica si ha pasado más de 24 horas desde la última sincronización.
    var needsSync: Bool {
        guard isEnabled else { return false }
        guard let last = lastSyncDate else { return true }
        return Date().timeIntervalSince(last) > 86400 // 24 horas
    }

    /// Versión del esquema de caché offline. Se incrementa cuando introducimos
    /// nuevos tipos de artefactos (docs R2, banderas, logos…) para que los
    /// usuarios que ya tenían el modo offline activo disparen un sync completo
    /// en la primera apertura tras actualizar la app — si no, tendrían datos
    /// incompletos hasta el siguiente sync periódico.
    ///
    /// Historial:
    ///   - 0 → inicial (solo JSON de días/meses/temporada).
    ///   - 1 → añade descarga de docs R2 (PDFs, mapas, perfiles…).
    ///   - 2 → añade descarga de banderas + logos de carrera.
    static let cacheSchemaVersion: Int = 2

    /// Versión guardada localmente; `0` si no existe (primer arranque o app
    /// pre-migración).
    var storedCacheSchemaVersion: Int {
        get { UserDefaults.standard.integer(forKey: "offline_cache_schema_version") }
        set { UserDefaults.standard.set(newValue, forKey: "offline_cache_schema_version") }
    }

    /// `true` si el usuario tiene offline activo y la caché local es de una
    /// versión previa — hay que forzar un sync para recoger los nuevos tipos
    /// de artefactos.
    var needsMigrationSync: Bool {
        isEnabled && storedCacheSchemaVersion < Self.cacheSchemaVersion
    }

    /// `true` si la fecha (formato `YYYY-MM-DD`) cae dentro de la ventana que
    /// el sync mantiene al día: mes en curso o mes siguiente. Los 14 días de
    /// la vista "Hoy" siempre caen dentro, así que no hace falta comprobarlos
    /// aparte. El sync de temporada solo descarga `races`, no `race_days`, por
    /// lo que NO lo consideramos "rango" para pull-to-refresh de jornadas.
    func isInOfflineRange(dateKey: String) -> Bool {
        let parts = dateKey.split(separator: "-")
        guard parts.count >= 2,
              let year = Int(parts[0]),
              let month = Int(parts[1]) else { return false }
        let cal = Calendar(identifier: .iso8601)
        let now = Date()
        let currentYear = cal.component(.year, from: now)
        let currentMonth = cal.component(.month, from: now)
        let nextMonth: Int
        let nextMonthYear: Int
        if currentMonth == 12 {
            nextMonth = 1
            nextMonthYear = currentYear + 1
        } else {
            nextMonth = currentMonth + 1
            nextMonthYear = currentYear
        }
        return (year == currentYear && month == currentMonth)
            || (year == nextMonthYear && month == nextMonth)
    }

    private init() {}

    // MARK: - API pública

    /// Activa el modo offline y realiza la descarga inicial.
    func enable() async {
        isEnabled = true
        await performSync()
    }

    /// Desactiva el modo offline y elimina los datos descargados.
    func disable() async {
        isEnabled = false
        lastSyncDate = nil
        syncProgress = 0
        syncStatusText = nil
        await CacheManager.shared.clearOfflineData()
    }

    /// Realiza la sincronización diaria: descarga datos y purga los obsoletos.
    /// Se llama automáticamente al abrir la app si ha pasado > 24h.
    func performSync() async {
        guard isEnabled, !isSyncing else { return }
        isSyncing = true
        syncProgress = 0

        let cache = CacheManager.shared
        let today = DateFormatting.todayKey()
        let cal = Calendar(identifier: .iso8601)
        let currentYear = cal.component(.year, from: Date())
        let currentMonth = cal.component(.month, from: Date())

        // Calcular mes siguiente
        let nextMonth: Int
        let nextMonthYear: Int
        if currentMonth == 12 {
            nextMonth = 1
            nextMonthYear = currentYear + 1
        } else {
            nextMonth = currentMonth + 1
            nextMonthYear = currentYear
        }

        // Total de pasos: 14 días + 2 meses + 1 temporada + 1 descarga R2 +
        // 1 descarga imágenes UI + 1 purga = 19
        let totalSteps = 19.0
        var completedSteps = 0.0

        /// IDs de assets R2 retenidos en esta pasada — se usan al final para
        /// purgar ficheros de jornadas que ya no están en la ventana offline.
        var retainedAssetIds = Set<String>()
        var assetsToDownload: [Asset] = []

        /// URLs remotas de logos de carrera a descargar (hash del absoluteString → nombre de fichero).
        var retainedLogoURLs = Set<URL>()

        do {
            // 1. Descargar los próximos 14 días (vista Hoy)
            syncStatusText = "Descargando agenda diaria…"
            for offset in 0..<14 {
                guard let dateKey = DateFormatting.dayOffset(from: today, by: offset) else { continue }
                let cacheKey = CacheManager.dayKey(dateKey)

                var dayData: DayData?

                // Solo descargar si no hay datos recientes (< 12h)
                if let age = await cache.age(forKey: cacheKey), age < 43200 {
                    dayData = await cache.load(DayData.self, forKey: cacheKey)
                    completedSteps += 1
                    syncProgress = completedSteps / totalSteps
                } else {
                    do {
                        let data = try await SupabaseService.shared.loadDayComplete(dateKey: dateKey)
                        await cache.save(data, forKey: cacheKey)
                        dayData = data
                    } catch {
                        #if DEBUG
                        print("[OfflineManager] Error sincronizando día \(dateKey): \(error.localizedDescription)")
                        #endif
                    }
                    completedSteps += 1
                    syncProgress = completedSteps / totalSteps
                }

                // Acumular los assets R2 de esta jornada (aunque vengan de caché,
                // hay que retenerlos para que no los purgue el paso final).
                if let data = dayData {
                    for enriched in data.raceDays {
                        for asset in enriched.assets where asset.isDownloadableR2 {
                            retainedAssetIds.insert(asset.id)
                            assetsToDownload.append(asset)
                        }
                    }
                    // Logos de todas las carreras del mapa (incluye placeholders y expandidas por jornada).
                    for race in data.raceMap.values {
                        Self.collectArtwork(from: race, logos: &retainedLogoURLs)
                    }
                }
            }

            // 2. Descargar mes actual (vista Mes)
            syncStatusText = "Descargando mes actual…"
            await downloadMonth(year: currentYear, month: currentMonth, cache: cache, logos: &retainedLogoURLs)
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            // 3. Descargar mes siguiente (vista Mes)
            syncStatusText = "Descargando mes siguiente…"
            await downloadMonth(year: nextMonthYear, month: nextMonth, cache: cache, logos: &retainedLogoURLs)
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            // 4. Descargar temporada completa (vista Temporada)
            syncStatusText = "Descargando temporada…"
            let seasonKey = CacheManager.seasonKey(currentYear)
            do {
                let races = try await SupabaseService.shared.racesByYear(currentYear)
                await cache.save(races, forKey: seasonKey)
                for race in races {
                    Self.collectArtwork(from: race, logos: &retainedLogoURLs)
                }
            } catch {
                #if DEBUG
                print("[OfflineManager] Error sincronizando temporada \(currentYear): \(error.localizedDescription)")
                #endif
            }
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            // 5. Descargar los ficheros R2 (PDFs, mapas, perfiles) de las jornadas
            //    cacheadas arriba. Esto hace que la documentación sea accesible
            //    sin conexión desde el detalle de etapa. URLs externas (live_text,
            //    TV oficial) se ignoran automáticamente vía `isDownloadableR2`.
            syncStatusText = "Descargando documentación… (puede tardar 1-2 min)"
            // Deduplicar por asset.id (la misma jornada aparece solo una vez,
            // pero un asset puede repetirse si se re-evaluó entre días).
            var seen = Set<String>()
            let uniqueAssets = assetsToDownload.filter { seen.insert($0.id).inserted }
            await cache.downloadAssets(uniqueAssets)
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            // 6. Descargar logos de carrera (R2) para los que aún no están en el
            //    bundle empaquetado (carreras nuevas añadidas tras el último build).
            //    Las banderas van ya en el asset catalog — no se descargan.
            syncStatusText = "Descargando logos…"
            await cache.downloadLogos(retainedLogoURLs)
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            // 7. Purgar datos obsoletos — JSON fuera de ventana + ficheros de
            //    assets cuyas jornadas ya no están cacheadas + imágenes UI de
            //    carreras que ya no están en la ventana.
            syncStatusText = "Limpiando datos antiguos…"
            await cache.purgeExpiredOfflineData(currentDateKey: today)
            await cache.purgeAssetFiles(keeping: retainedAssetIds)
            await cache.purgeImages(keepingLogoURLs: retainedLogoURLs)
            completedSteps += 1
            syncProgress = completedSteps / totalSteps

            lastSyncDate = Date()
            // Marca el esquema de caché al día — los próximos arranques ya no
            // disparan la migración.
            storedCacheSchemaVersion = Self.cacheSchemaVersion
            syncStatusText = nil

            // Actualizar widget con los datos recién sincronizados de hoy
            let todayKey = DateFormatting.todayKey()
            if let dayData: DayData = await CacheManager.shared.load(
                DayData.self, forKey: CacheManager.dayKey(todayKey)
            ) {
                writeWidgetPayload(items: dayData.raceDays, nextDateKey: nil, dateKey: todayKey)
            }
        }

        isSyncing = false
        syncProgress = 1.0
    }

    /// Comprueba si es necesario sincronizar y lo hace automáticamente.
    /// Llamar al iniciar la app.
    ///
    /// Dispara sync si toca por tiempo (>24 h) o por migración de esquema
    /// (nuevos tipos de artefactos añadidos a la caché offline).
    func syncIfNeeded() async {
        guard needsSync || needsMigrationSync else { return }
        await performSync()
    }

    // MARK: - Privado

    /// Descarga los datos de un mes completo para la vista de Mes.
    /// Acumula en `logos` las URLs de logos a descargar en el paso 6 del sync.
    private func downloadMonth(year: Int,
                               month: Int,
                               cache: CacheManager,
                               logos: inout Set<URL>) async {
        let cacheKey = CacheManager.monthKey(year: year, month: month)
        let needsFetch: Bool
        if let age = await cache.age(forKey: cacheKey), age < 43200 {
            needsFetch = false
        } else {
            needsFetch = true
        }

        // Estructura compatible con MonthViewModel.MonthCache.
        struct MonthCache: Codable {
            let raceDays: [RaceDay]
            let races: [Race]
        }

        if !needsFetch {
            // Incluso con caché fresca, hay que recolectar logos de sus
            // carreras para retenerlos durante la purga.
            if let existing = await cache.load(MonthCache.self, forKey: cacheKey) {
                for race in existing.races {
                    Self.collectArtwork(from: race, logos: &logos)
                }
            }
            return
        }

        let neededMonth = String(format: "%04d-%02d", year, month)
        let cal = Calendar(identifier: .iso8601)
        let firstOfMonth = cal.date(from: DateComponents(year: year, month: month, day: 1)) ?? Date()
        let daysInMonth = cal.range(of: .day, in: .month, for: firstOfMonth)?.count ?? 30
        let startKey = "\(neededMonth)-01"
        let endKey = String(format: "%@-%02d", neededMonth, daysInMonth)

        do {
            async let daysResult = SupabaseService.shared.raceDays(from: startKey, to: endKey)
            async let racesResult = SupabaseService.shared.racesByYear(year)

            let (publishedDays, races) = try await (daysResult, racesResult)

            await cache.save(MonthCache(raceDays: publishedDays, races: races), forKey: cacheKey)

            // También cachear las carreras del año
            let yearKey = CacheManager.yearRacesKey(year)
            await cache.save(races, forKey: yearKey)

            for race in races {
                Self.collectArtwork(from: race, logos: &logos)
            }
        } catch {
            #if DEBUG
            print("[OfflineManager] Error sincronizando mes \(month)/\(year): \(error.localizedDescription)")
            #endif
        }
    }

    /// Extrae la URL del logo de una carrera hacia el set acumulador.
    /// Las banderas van en el asset catalog — no se gestionan aquí.
    static func collectArtwork(from race: Race, logos: inout Set<URL>) {
        if let logo = race.logoUrl, let url = URL(string: logo) {
            logos.insert(url)
        }
    }
}

File: android-app/app/src/main/java/app/calendariociclismo/android/data/sync/OfflineManager.kt
package app.calendariociclismo.android.data.sync

import android.content.Context
import android.util.Log
import androidx.glance.appwidget.updateAll
import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import app.calendariociclismo.android.data.model.Asset
import app.calendariociclismo.android.data.model.Race
import app.calendariociclismo.android.data.prefs.AppPreferences
import app.calendariociclismo.android.data.repository.CalendarRepository
import app.calendariociclismo.android.util.DateFormatting
import app.calendariociclismo.android.widget.today.TodayCyclingWidget
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.LocalDate
import java.time.YearMonth
import java.util.concurrent.TimeUnit

/**
 * Coordinador del modo offline.
 *
 * Port de `OfflineManager.swift`. Ajustes:
 *   - Persistencia de estado vía [AppPreferences] (DataStore).
 *   - Caché local vía [CalendarRepository] (Room).
 *   - Sync en background vía [OfflineSyncWorker] (WorkManager).
 *   - Estado observable como `StateFlow<SyncState>`.
 *
 * La estrategia replica la iOS:
 *   1. Próximos 14 días (vista Hoy)
 *   2. Mes actual completo (vista Mes)
 *   3. Mes siguiente completo (vista Mes)
 *   4. Temporada del año en curso (vista Temporada)
 *   5. Purga de filas con cachedAt muy antiguo
 *
 * Total de pasos: 14 + 2 + 1 = 17.
 */
class OfflineManager(
    private val appContext: Context,
    private val prefs: AppPreferences,
    private val repo: CalendarRepository,
    private val assetCache: FileAssetCache = FileAssetCache(appContext),
    private val imageCache: ImageAssetCache = ImageAssetCache(appContext),
) {
    private val TAG = "OfflineManager"
    private val syncMutex = Mutex()

    private val _state = MutableStateFlow(SyncState())
    val state: StateFlow<SyncState> = _state.asStateFlow()

    // ─── API pública ───

    suspend fun enable() {
        prefs.setOfflineEnabled(true)
        schedulePeriodic()
        runSyncNow(force = true)
    }

    suspend fun disable() {
        prefs.setOfflineEnabled(false)
        cancelPeriodic()
        repo.clearAll()
        runCatching { assetCache.clear() }
            .onFailure { Log.w(TAG, "Error limpiando caché de assets: ${it.message}") }
        runCatching { imageCache.clear() }
            .onFailure { Log.w(TAG, "Error limpiando caché de imágenes: ${it.message}") }
        _state.value = SyncState()
        prefs.recordSync(0L, "cleared")
    }

    /** Acceso a la caché de ficheros R2 (usado desde la UI de detalle de etapa). */
    fun assetCache(): FileAssetCache = assetCache

    /**
     * `true` si la fecha (formato `YYYY-MM-DD`) cae dentro de la ventana que
     * el sync mantiene al día: mes en curso o mes siguiente. Los 14 días de
     * la vista "Hoy" siempre caen dentro, así que no hace falta comprobarlos
     * aparte. El sync de temporada solo descarga `races`, no `race_days`, por
     * lo que NO lo consideramos "rango" para pull-to-refresh de jornadas.
     */
    fun isInOfflineRange(dateKey: String): Boolean {
        val parts = dateKey.split("-")
        if (parts.size < 2) return false
        val year = parts[0].toIntOrNull() ?: return false
        val month = parts[1].toIntOrNull() ?: return false
        val today = LocalDate.now()
        val current = YearMonth.of(today.year, today.monthValue)
        val next = current.plusMonths(1)
        val ym = runCatching { YearMonth.of(year, month) }.getOrNull() ?: return false
        return ym == current || ym == next
    }

    /**
     * Lanza un sync one-shot. [force] solo se usa para acciones explícitas del
     * usuario; los arranques y reintentos automáticos respetan el cooldown.
     */
    fun runSyncNow(force: Boolean = false) {
        val req = OneTimeWorkRequestBuilder<OfflineSyncWorker>()
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            )
            .setBackoffCriteria(BackoffPolicy.EXPONENTIAL, 30, TimeUnit.SECONDS)
            .setInputData(workDataOf(KEY_FORCE_SYNC to force))
            .build()
        WorkManager.getInstance(appContext).enqueueUniqueWork(
            WORK_ONESHOT,
            // Los disparos automáticos no reinician un trabajo activo. Una
            // acción explícita sí sustituye una cola/backoff pendiente para
            // que «Actualizar ahora» nunca quede absorbido por KEEP.
            oneShotPolicy(force),
            req,
        )
    }

    /** Registra el worker periódico de 24 h con red no medida. */
    fun schedulePeriodic() {
        val req = PeriodicWorkRequestBuilder<OfflineSyncWorker>(24, TimeUnit.HOURS)
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.UNMETERED)
                    .setRequiresBatteryNotLow(true)
                    .build()
            )
            .setBackoffCriteria(BackoffPolicy.EXPONENTIAL, 1, TimeUnit.HOURS)
            .build()
        WorkManager.getInstance(appContext).enqueueUniquePeriodicWork(
            WORK_PERIODIC,
            ExistingPeriodicWorkPolicy.KEEP,
            req,
        )
    }

    fun cancelPeriodic() {
        WorkManager.getInstance(appContext).cancelUniqueWork(WORK_PERIODIC)
    }

    /**
     * Ejecución real de la sincronización. Llamada desde [OfflineSyncWorker].
     *
     * Protegida por mutex para evitar ejecuciones concurrentes si el usuario
     * dispara un one-shot mientras el periódico está corriendo.
     */
    suspend fun performSync(force: Boolean = false): Result<Unit> = syncMutex.withLock {
        if (!prefs.snapshotOfflineEnabled()) {
            Log.i(TAG, "Offline deshabilitado, saltando sync")
            return@withLock Result.success(Unit)
        }

        val attemptEpoch = System.currentTimeMillis() / 1000
        val lastAttemptEpoch = prefs.snapshotLastOfflineSyncAttemptEpoch()
        if (!force && OfflineSyncThrottle.shouldSkip(attemptEpoch, lastAttemptEpoch)) {
            Log.i(
                TAG,
                "Sync automática omitida por cooldown " +
                    "(último intento hace ${attemptEpoch - lastAttemptEpoch}s)",
            )
            return@withLock Result.success(Unit)
        }
        // Persistir antes de la primera petición: también protege si Android
        // mata el proceso durante la descarga.
        prefs.setLastOfflineSyncAttemptEpoch(attemptEpoch)

        _state.value = _state.value.copy(isSyncing = true, progress = 0f, lastError = null)
        // 14 días + 2 meses + 1 temporada + 1 descarga R2 + 1 descarga imágenes +
        // 1 purga = 19
        val totalSteps = 19f
        var completed = 0f

        try {
            val todayKey = DateFormatting.todayKey()
            val today = LocalDate.now()
            val currentYear = today.year
            val currentMonth = YearMonth.from(today)
            val nextMonth = currentMonth.plusMonths(1)

            // IDs de jornadas cacheadas esta pasada (para recoger sus assets R2
            // de Room al final y purgar los ficheros huérfanos).
            val syncedRaceDayIds = mutableSetOf<String>()

            // 1. Próximos 14 días
            setStatus("Descargando agenda diaria…")
            for (offset in 0 until 14) {
                val dateKey = DateFormatting.dayOffset(todayKey, offset) ?: continue
                runCatching { repo.refreshDay(dateKey) }
                    .onFailure { Log.w(TAG, "Error día $dateKey: ${it.message}") }
                // Recoger los IDs de jornadas ahora cacheadas — incluso si la
                // llamada a refreshDay falló, puede haber datos previos válidos.
                runCatching {
                    val ids = repo.cachedRaceDaysByDate(dateKey).map { it.id }
                    syncedRaceDayIds.addAll(ids)
                }
                completed++
                publishProgress(completed / totalSteps)
            }

            // 2. Mes actual
            setStatus("Descargando mes actual…")
            downloadMonth(currentMonth)
            completed++
            publishProgress(completed / totalSteps)

            // 3. Mes siguiente
            setStatus("Descargando mes siguiente…")
            downloadMonth(nextMonth)
            completed++
            publishProgress(completed / totalSteps)

            // 4. Temporada
            setStatus("Descargando temporada…")
            runCatching { repo.refreshRacesYear(currentYear) }
                .onFailure { Log.w(TAG, "Error temporada $currentYear: ${it.message}") }
            completed++
            publishProgress(completed / totalSteps)

            // 5. Descargar ficheros R2 (PDFs, mapas, perfiles) de las jornadas
            //    cacheadas arriba — misma idea que iOS. URLs externas se
            //    descartan automáticamente (filtrado por isDownloadableR2).
            setStatus("Descargando documentación… (puede tardar 1-2 min)")
            val retainedAssetIds = mutableSetOf<String>()
            val assetsToDownload = mutableListOf<Asset>()
            val seen = HashSet<String>()
            runCatching {
                val assets = repo.cachedAssetsForRaceDays(syncedRaceDayIds.toList())
                for (asset in assets) {
                    if (!asset.isDownloadableR2) continue
                    retainedAssetIds.add(asset.id)
                    if (seen.add(asset.id)) assetsToDownload.add(asset)
                }
            }.onFailure { Log.w(TAG, "Error listando assets: ${it.message}") }
            if (assetsToDownload.isNotEmpty()) {
                runCatching { assetCache.downloadAll(assetsToDownload) }
                    .onFailure { Log.w(TAG, "Error descarga R2: ${it.message}") }
            }
            completed++
            publishProgress(completed / totalSteps)

            // 6. Descargar logos de carrera (R2) para los que aún no están en el
            //    bundle empaquetado (carreras nuevas añadidas tras el último build).
            //    Las banderas van en assets/flags/ (bundled) — no se descargan.
            setStatus("Descargando logos…")
            val retainedLogos = hashSetOf<String>()
            runCatching {
                val racesForYear = repo.cachedRacesForYear(currentYear)
                for (r in racesForYear) collectArtwork(r, retainedLogos)
                // Carreras referenciadas por jornadas cacheadas — incluye
                // cualquier carrera que no esté en `currentYear` (ej: Grandes
                // Vueltas que cruzan 2 años) y también placeholders.
                val raceIds = repo.cachedRaceIdsForRaceDays(syncedRaceDayIds.toList())
                for (id in raceIds) repo.cachedRace(id)?.let { collectArtwork(it, retainedLogos) }
            }.onFailure { Log.w(TAG, "Error recogiendo logos: ${it.message}") }
            if (retainedLogos.isNotEmpty()) {
                runCatching { imageCache.downloadLogos(retainedLogos) }
                    .onFailure { Log.w(TAG, "Error descarga logos: ${it.message}") }
            }
            completed++
            publishProgress(completed / totalSteps)

            // 7. Purga de filas viejas (>21 días → limpio) + ficheros R2 huérfanos
            //    + imágenes UI de carreras fuera de la ventana.
            setStatus("Limpiando datos antiguos…")
            val cutoff = (System.currentTimeMillis() / 1000) - 21L * 86400
            repo.purgeStale(cutoff)
            runCatching { assetCache.purge(retainedAssetIds) }
                .onFailure { Log.w(TAG, "Error purgando assets: ${it.message}") }
            runCatching { imageCache.purge(retainedLogos) }
                .onFailure { Log.w(TAG, "Error purgando imágenes: ${it.message}") }

            val now = System.currentTimeMillis() / 1000
            prefs.recordSync(now, "ok")
            // Marca el esquema de caché al día — los próximos arranques ya no
            // disparan la migración. Ver [CACHE_SCHEMA_VERSION] para historial.
            prefs.setOfflineCacheSchemaVersion(CACHE_SCHEMA_VERSION)
            // Redibujar el widget con los datos recién sincronizados
            runCatching { TodayCyclingWidget().updateAll(appContext) }
                .onFailure { Log.w(TAG, "Error actualizando widget tras sync: ${it.message}") }
            _state.value = SyncState(
                isSyncing = false,
                progress = 1f,
                statusText = null,
                lastSyncEpochSeconds = now,
                lastError = null,
            )
            return@withLock Result.success(Unit)
        } catch (t: Throwable) {
            Log.e(TAG, "Sync fallida", t)
            _state.value = _state.value.copy(
                isSyncing = false,
                statusText = null,
                lastError = t.message ?: "Error desconocido",
            )
            prefs.recordSync(System.currentTimeMillis() / 1000, "error")
            return@withLock Result.failure(t)
        }
    }

    private suspend fun downloadMonth(month: YearMonth) {
        val startKey = "%04d-%02d-01".format(month.year, month.monthValue)
        val endKey = "%04d-%02d-%02d".format(month.year, month.monthValue, month.lengthOfMonth())
        runCatching {
            repo.refreshRange(startKey, endKey)
            repo.refreshRacesYear(month.year)
        }.onFailure { Log.w(TAG, "Error mes ${month.year}-${month.monthValue}: ${it.message}") }
    }

    /** Extrae la URL del logo de una carrera hacia el set acumulador. */
    private fun collectArtwork(race: Race, logos: MutableSet<String>) {
        race.logoUrl?.takeIf { it.isNotEmpty() }?.let(logos::add)
    }

    private fun publishProgress(value: Float) {
        _state.value = _state.value.copy(progress = value.coerceIn(0f, 1f))
    }

    private fun setStatus(text: String) {
        _state.value = _state.value.copy(statusText = text)
    }

    companion object {
        const val WORK_PERIODIC = "offline_sync_periodic"
        const val WORK_ONESHOT = "offline_sync_oneshot"
        const val KEY_FORCE_SYNC = "force_sync"

        internal fun oneShotPolicy(force: Boolean): ExistingWorkPolicy =
            if (force) ExistingWorkPolicy.REPLACE else ExistingWorkPolicy.KEEP

        /**
         * Versión del esquema de caché offline. Se incrementa cuando añadimos
         * tipos de artefactos nuevos para que los usuarios que ya tenían el
         * modo offline activo disparen un sync completo en la primera apertura
         * tras actualizar la app — si no, tendrían datos incompletos hasta el
         * siguiente sync periódico (hasta 24 h).
         *
         * Historial:
         *   - 0 → inicial (solo JSON de días / meses / temporada).
         *   - 1 → añade descarga de docs R2 (PDFs, mapas, perfiles…).
         *   - 2 → añade descarga de banderas + logos de carrera.
         *   - 3 → logos bundled en assets; flags siempre bundled → se elimina
         *         la descarga de banderas del sync. Solo se descargan logos de
         *         carreras nuevas no incluidas en el bundle empaquetado.
         */
        const val CACHE_SCHEMA_VERSION = 3
    }
}
